feat: Add Brew module (#239)
blue-build/modules: modules/brew/README.md

> # brew

Reviewer: I just remembered that `brew.sh` is included in `ublue-os-just` RPM package.
It resides in `/usr/etc/profile.d/brew.sh`

It adds brew in path only when shell is interactive, solving issues with `brew` path clashing with other programs

```bash
#!/usr/bin/env bash
[[ -d /home/linuxbrew/.linuxbrew && $- == *i* ]] && eval "$(/home/linuxbrew/.linuxbrew/bin/brew shellenv)"
```

fish already contains this part in `brew-fish-completions`.

This is only in Universal Blue images.

We need to account for other images which don't include this.

> The brew module installs [Homebrew (Brew)](https://brew.sh/) on your system and ensures the package manager remains updated and maintained. This module sets up systemd services to periodically update and upgrade the installed Brew packages.

Reviewer 2: ```suggestion
The brew module installs [Homebrew/Linuxbrew](https://brew.sh/) on your system and ensures the package manager remains updated and maintained. This module also sets up systemd services to periodically update the installed Brew packages.
```

> ## Features
> - Installs Brew at build-time.
> - Sets up systemd services to update the Brew binary to the latest version.
> - Sets up systemd services to upgrade Brew packages automatically.
> - Options to control the frequency of updates and upgrades.

Reviewer 2: ```suggestion
## Features
- Installs Brew at build-time.
- Sets up systemd services to automatically update Brew to the latest version.
- Sets up systemd services to automatically upgrade Brew packages.
- Sets up bash and fish completions for Brew.
```

> ## How it works

Reviewer 2: All these implementation details are great and I enjoyed reading it. However, is it more important than the available configuration options? I'm not sure how I would solve this problem, though, since I have [an open issue](https://github.com/blue-build/website/issues/55) for generating the configuration options -section from the schema. Maybe it's not urgent right now and I can solve it later...

Reviewer: Maybe I can move the section down below. I thought about it, but forgot it.

Reviewer 2: I'll mark this as resolved, it's not that important.

  - `/var/lib/homebrew/`
  - `/var/cache/homebrew/`
  - `/home/linuxbrew/`
- `brew-setup` service is enabled

### Boot-time:

**`tmpfiles.d homebrew.conf`:**
- This configuration is telling SystemD to: automatically create these necessary directories on every system boot if not available & to give them permissions of the default user (UID 1000):
  - `/var/lib/homebrew/`
  - `/var/cache/homebrew/`
  - `/home/linuxbrew/`

**`brew-setup`:**
- `brew-setup` SystemD service checks if main directory used by Brew exists (`/home/linuxbrew/.linuxbrew/`)  
  & if `brew-setup` state file exists (`/etc/.linuxbrew`)
- If one of those paths don't exist, than Homebrew tar is extracted from `/usr/share/homebrew/homebrew.tar.zst` to `/tmp/homebrew/`
- Extracted Homebrew is than copied from `/tmp/homebrew/` to `/home/linuxbrew/` & permissions to it are set to default user (UID 1000)
- Temporary directory `/tmp/homebrew/` is removed
- Empty file `/etc/.linuxbrew` is created, which indicates that brew-setup (installation) is successful & which allows setup to run again on next boot when removed

**Rest of the setup:**

> - `brew-update` is runnning in specified time to update Brew binary to latest version
> - `brew-upgrade` is runnning in specified time to upgrade Brew packages to latest versions

Reviewer 2: ```suggestion
- `brew-update` is runs at the specified time to update Brew to the latest version
- `brew-upgrade` is runs at the specified time to upgrade Brew packages```

## Configuration Options

### `update-interval` (optional: string, default: '6h')
Defines how often the Brew update service should run. The string is passed directly to `OnUnitInactiveSec` in systemd timer. (Syntax: ['1d', '6h', '10m']).

### `upgrade-interval` (optional: string, default: '8h')
Defines how often the Brew upgrade service should run. The string is passed directly to `OnUnitInactiveSec` in systemd timer. (Syntax: ['1d', '6h', '10m']).

### `auto-update` (optional: boolean, default: true)
If false, disables automatic activation of `brew-update.timer`.

### `wait-after-boot-update` (optional: string, default: '10min')
Time delay after system boot before the first Brew update runs. The string is passed directly to `OnBootSec` in systemd timer. (Syntax: ['1d', '6h', '10m']).

### `auto-upgrade` (optional: boolean, default: true)
If false, disables automatic activation of `brew-upgrade.timer`.

### `wait-after-boot-upgrade` (optional: string, default: '30min')

> Time delay after system boot before the first Brew package upgrade runs. The string is passed directly to `OnBootSec` in systemd timer. (Syntax: ['1d', '6h', '10m']).

Reviewer 2: Ok, how about this ordering:
- `auto-update`
   - need to add definition of `update` to distinguish from `upgrade`
- `update-interval`
- `wait-after-boot-update`
   - feels like this could be renamed to `update-wait-after-boot`
- `auto-upgrade`
   - need to add definition of `upgrade` to distinguish from `update`
- `upgrade-interval`
- `wait-after-boot-upgrade`
   - feels like this could be renamed to `upgrade-wait-after-boot`

Reviewer: I agree with this.

I will change the code to also reflect this.

Reviewer: About

`need to add definition of upgrade to distinguish from update`

Idk what would be the best suggestion.

> If false, disables automatic activation of `brew-upgrade.timer`.

Reviewer 2: Move these above the wait timers, they will probably be the most sought after.

!!! warning

>     Please review the brew documentation carefully before modifying these settings.

Reviewer 2: ```suggestion
    Please review the Brew documentation carefully before modifying the settings below.
```

### `nofile-limits` (optional: boolean, default: false)

> Determines whether to apply nofile limits for Brew installations. When set to true, it increases the nofile limits to prevent certain packages from failing due to file limits. However, it's important to note that increasing nofile limits can have potential security implications. Defaults to false for security purposes.

Reviewer 2: I have no idea what `nofile limits` and what that would mean for me. I'd also be interested in knowing if these limits are set by default when installing brew the official way. I googled the term, but did not find any information about security implications. 

I also assume that the warning above applies to this setting mostly, so a link to the referenced Brew documentation would be great, if that exists.

Reviewer: It is not set by default in `brew`.

Homebrew doesn't have documentation about this either, nor they are not aware about this afaik.

It is something that Bluefin maintainers noticed, where some packages installed from `brew` complain about reached nofile limits & refuse to work due to this.
https://github.com/ublue-os/bluefin/pull/988

This doesn't just happen with `brew` packages in particular, it can happen with other packages too.
Why this happens in more detail, I have no idea.

About security implications, this is just my own view of it. It is indeed true that there are no searchable topics regarding security of nofile limits.

Maybe @qoijjj knows something more about this.

### `brew-analytics` (optional: boolean, default: true)

> Determines whether to opt-out of Brew analytics. When set to true, analytics are enabled.

Reviewer 2: Is this in scope for the warning above? If not, consider moving it above the warning. IMO analytics opt-out is more user preference than something that should be warned about.

Reviewer: Yes, I can move this.

## Development
Setting `DEBUG=true` inside `brew.sh` will enable additional output for debugging purposes during development.

## Uninstallation

> When excluding `brew` module from the recipe, it's not enough to get it removed.  
> On booted system, it's also necessary to run the `brew` uninstalation script.
> 
> Either local-user can execute this script manually or image-maintainer can make it automatic through SystemD service.

Reviewer 2: ```suggestion
Removing the `brew` module from the recipe is not enough to get it completely removed.   
On a booted system, it's also necessary to run the `brew` uninstallation script.

Either a local-user can execute this script manually or the image-maintainer may make it automatic through a custom systemd service.
```

Uninstallation script:  

> ```

Reviewer 2: ```suggestion
```bash
```

(syntax highlighting)

**Rest of the setup:**
- `brew-update` runs at the specified time to update Brew to the latest version
- `brew-upgrade` runs at the specified time to upgrade Brew packages

## Configuration Options

### `auto-update` (optional: boolean, default: true)
If false, disables automatic activation of `brew-update.timer`.

### `update-interval` (optional: string, default: '6h')
Defines how often the Brew update service should run. The string is passed directly to `OnUnitInactiveSec` in systemd timer. (Syntax: ['1d', '6h', '10m']).

### `update-wait-after-boot` (optional: string, default: '10min')
Time delay after system boot before the first Brew update runs. The string is passed directly to `OnBootSec` in systemd timer. (Syntax: ['1d', '6h', '10m']).

### `auto-upgrade` (optional: boolean, default: true)
If false, disables automatic activation of `brew-upgrade.timer`.

### `upgrade-interval` (optional: string, default: '8h')
Defines how often the Brew upgrade service should run. The string is passed directly to `OnUnitInactiveSec` in systemd timer. (Syntax: ['1d', '6h', '10m']).

### `upgrade-wait-after-boot` (optional: string, default: '30min')

> Time delay after system boot before the first Brew package upgrade runs. The string is passed directly to `OnBootSec` in systemd timer. (Syntax: ['1d', '6h', '10m']).

Reviewer 2: Somewhere around here the difference between updates and upgrades should be explained.

Reviewer: I got the better idea. Please check out this commit:

[`8ad2c3e` (#239)](https://github.com/blue-build/modules/pull/239/commits/8ad2c3ec05e2b401466186e36c101e9051c811dd)

# brew

The brew module installs Homebrew (Brew) on your system and ensures the package manager remains updated and maintained. This module sets up systemd services to periodically update and upgrade the installed Brew packages.

> ## Features

Reviewer 3: That's a lot of whitespace. Mind cleaning that up?

- Installs Brew at build time.

- Configures and installs specified Brew packages.

- Sets up systemd services to update and upgrade Brew packages automatically.

- Options to control the frequency of updates and upgrades.

## Configuration Options

> ### `install_brew` (optional: boolean, default: true)

Reviewer 2: What is the module used for if not installing brew? How do we expect the services to add brew packages and update brew to work if brew isn't installed?

Is there a possibility of conflict if a user has brew installed prior to switching to an image with brew pre-installed, which would make not pre-installing brew while using the module a valid choice? If so, we should document that very clearly as a warning.

IMO (and with the limited information I have), just including `- type: brew` should be enough to install brew, such as including `- type: yafti` installs yafti.

Author: > What is the module used for if not installing brew? How do we expect the services to add brew packages and update brew to work if brew isn't installed?
> 
> Is there a possibility of conflict if a user has brew installed prior to switching to an image with brew pre-installed, which would make not pre-installing brew while using the module a valid choice? If so, we should document that very clearly as a warning.
> 
I drew inspiration from the [chezmoi module](https://github.com/blue-build/modules/blob/main/modules/chezmoi/module.yml#L7), which does have an "install" boolean. 

> IMO (and with the limited information I have), just including `- type: brew` should be enough to install brew, such as including `- type: yafti` installs yafti.

I agree, will remove that option :+1:

Reviewer 2: > I drew inspiration from the chezmoi module, which does have an "install" boolean.

That module has some issues too, which unfortunately made it through review. Trying to catch those before merge in the future.

Determines whether Brew should be installed at build time. If set to `false`, Brew will not be installed.

### `packages` (required: list of strings)

> A list of Brew packages to be installed. This is a mandatory configuration and must be provided.

Reviewer 3: Should this be a requirement? I think we should allow the user to decide to install something at build time vs runtime.

Reviewer 2: +1

### `update_interval` (optional: string, default: '6h')

Defines how often the Brew update service should run. The string is passed directly to `OnUnitInactiveSec` in systemd timer. (Syntax: ['1d', '6h', '10m']).

### `upgrade_interval` (optional: string, default: '8h')

Defines how often the Brew upgrade service should run. The string is passed directly to `OnUnitInactiveSec` in systemd timer. (Syntax: ['1d', '6h', '10m']).

### `wait_after_boot_update` (optional: string, default: '10min')

Time delay after system boot before the first Brew update runs. The string is passed directly to `OnBootSec` in systemd timer. (Syntax: ['1d', '6h', '10m']).

### `wait_after_boot_upgrade` (optional: string, default: '30min')

Time delay after system boot before the first Brew package upgrade runs. The string is passed directly to `OnBootSec` in systemd timer. (Syntax: ['1d', '6h', '10m']).

> ### `disable_update` (optional: boolean, default: false)

Reviewer 2: This and `disable_upgrade` could be `auto-update` and `auto-upgrade`, which would be `true` by default and could be set to `false` by the user. That would be more clear IMO.

If set to `true`, disables the automatic activation of the `brew-update.timer`.

### `disable_upgrade` (optional: boolean, default: false)

If set to `true`, disables the automatic activation of the `brew-upgrade.timer`.

> ## Example Configuration

Reviewer 2: The example configuration is automatically added to the website documentation / reference page from `module.yml` and shouldn't be added to the README

> # brew

Reviewer: # Nofile limits

I think that we should document the necessary tweak to install/run some brew packages. Else, some packages complain about file limits & fail to run.

It is necessary to increase nofile limits, like it's done here (both limits.d & systemd config should be modified):

- SSH/tty nofile limit
`/usr/etc/security/limits.d/30-brew-limits.conf`

```
#This file sets the resource limits for the users logged in via PAM,
#more specifically, users logged in on via SSH or tty (console).
#Limits related to terminals in Wayland/Xorg sessions depend on a
#change to /etc/systemd/user.conf.
#This does not affect resource limits of the system services.
#This file overrides defaults set in /etc/security/limits.conf

* soft nofile 4096
* hard nofile 524288
```

- SystemD nofile limit affecting whole session
`/usr/lib/systemd/system/system.conf.d/30-brew-limits.conf`
`/usr/lib/systemd/user/user.conf.d/30-brew-limits.conf`

```
[Manager]
DefaultLimitNOFILE=4096:524288
```

# Necessary dependencies

Document that GCC needs to be installed in the image for brew to work.
I read that brew can provide gcc too, but I would let OS package manager to handle this one, since GCC is available in basically all distros.

Author: Should we add those overrides as lines of `cat` in brew.sh, the same way we did for the systemd units? 

> # Necessary dependencies
> Document that GCC needs to be installed in the image for brew to work. I read that brew can provide gcc too, but I would let OS package manager to handle this one, since GCC is available in basically all distros.

Recommended way: 
```
rpm:
  install:
    - gcc
```

Reviewer: I'm not sure if we should automatically add those for all users, since increasing nofile limits can have some potential security implications if application chooses to abuse it for evil purposes. I'm not the security expert, so can't explain it in detail, but that's how I see it.

It can also potentially override the same modification that someone did differently in their image for some other purposes.

I think that just recommendation in docs is better.

About GCC, I think that we can check if `gcc` command exists. If it doesn't, than inform the user that `gcc` is not installed, how it must be installed for brew to function & fail the build. We wouldn't add the else statement to do `rpm-ostree install`, since we try to make module to be universally compatible regarding container-based OSes (like Fedora Atomic & VanillaOS in the future)
Of course, we would still document this `gcc` requirement in docs.

Author: I think these two points can be resolved by improving brew.sh.
We could a flag that manages nofile-limits? It would default to False, and clearly mention the potential security implications if True.

Reviewer: I'll let others also discuss this review discussion & than try together to find the sweet-spot solution.

The brew module installs Homebrew (Brew) on your system and ensures the package manager remains updated and maintained. This module sets up systemd services to periodically update and upgrade the installed Brew packages.

## Features
- Installs Brew at build time.
- Configures and installs specified Brew packages.

> - Sets up systemd services to update and upgrade Brew packages automatically.

Reviewer: It should be more clearly noted about what is the difference between `update` & `upgrade` in README.

Most `update` & `upgrade` commands are the same with aliases, but that is not the case here.

`update` updates the Brew binary to latest version, while `upgrade` updates the installed Brew packages to latest version.

- Options to control the frequency of updates and upgrades.

## Configuration Options

### `packages` (required: list of strings)
A list of Brew packages to be installed. This is a mandatory configuration and must be provided.

### `update_interval` (optional: string, default: '6h')
Defines how often the Brew update service should run. The string is passed directly to `OnUnitInactiveSec` in systemd timer. (Syntax: ['1d', '6h', '10m']).

### `upgrade_interval` (optional: string, default: '8h')
Defines how often the Brew upgrade service should run. The string is passed directly to `OnUnitInactiveSec` in systemd timer. (Syntax: ['1d', '6h', '10m']).

### `wait_after_boot_update` (optional: string, default: '10min')
Time delay after system boot before the first Brew update runs. The string is passed directly to `OnBootSec` in systemd timer. (Syntax: ['1d', '6h', '10m']).

### `wait_after_boot_upgrade` (optional: string, default: '30min')
Time delay after system boot before the first Brew package upgrade runs. The string is passed directly to `OnBootSec` in systemd timer. (Syntax: ['1d', '6h', '10m']).

> ### `auto_update` (optional: boolean, default: false)

Reviewer: `auto-update` & `auto-upgrade` should be updated to contain `default: true` in README

If true, disables automatic activation of `brew-update.timer`.

### `auto_upgrade` (optional: boolean, default: false)
If true, disables automatic activation of `brew-upgrade.timer`.

## Development

> 

Reviewer 4: sorta a nitpick again, but the whitespace needs fixing here

Time delay after system boot before the first Brew update runs. The string is passed directly to `OnBootSec` in systemd timer. (Syntax: ['1d', '6h', '10m']).

### `wait-after-boot-upgrade` (optional: string, default: '30min')
Time delay after system boot before the first Brew package upgrade runs. The string is passed directly to `OnBootSec` in systemd timer. (Syntax: ['1d', '6h', '10m']).

### `auto-update` (optional: boolean, default: true)
If false, disables automatic activation of `brew-update.timer`.

### `auto-upgrade` (optional: boolean, default: true)
If false, disables automatic activation of `brew-upgrade.timer`.

!!! warning
    Please review the brew documentation carefully before modifying these settings.

### `nofile-limits` (optional: boolean, default: false)
Determines whether to apply nofile limits for Brew installations. When set to true, it increases the nofile limits to prevent certain packages from failing due to file limits. However, it's important to note that increasing nofile limits can have potential security implications. Defaults to false for security purposes.

### `brew-analytics` (optional: boolean, default: true)

> Determines whether to opt-out of Brew analytics. When set to true, analytics are disabled.

Reviewer 2: "When set to *true* analytics are *disabled*."

What? Mistake in the docs perhaps?

- Brew is extracted to `/home/linuxbrew/` by the official script (`/root/` is needed, since image-builds are running as root)
- Brew in `/home/linuxbrew/` is compressed in tar, copied to `/usr/share/homebrew/` & permissions to it are set to default user (UID 1000)
- `brew-update` & `brew-upgrade` SystemD service timers are enabled (by default)

> - Applied fix for program path conflict between system & brew packages with same name  
>   by exporting Brew path to shell environment when interactive shell is used only, unlike offical Homebrew way of exporting it globally (`/etc/profile.d/brew.sh`)

Reviewer 2: ```suggestion
- A fix for path conflicts between system & brew packages with the same name is applied by adding Brew to path only in interactive shells, unlike what Brew does by default.
```
